### ISSUE TITLE
Ed25519 accepting invalid signatures

### ISSUE THREAD

**Issue by paulidale** [openssl/openssl#13284]
According to this [blog post](https://hdevalence.ca/blog/2020-10-04-its-25519am), the OpenSSL Ed25519 implementation is accepting signatures that aren't valid.

The explanation given is:
> These implementations use the unbatched verification equation, and do not check that A is canonically encoded. However, because they recompute R’, encode it to Rprime_bytes, and check that R_bytes = Rprime_bytes, they do check that 
R is canonically encoded, but only by accident. The visualization reflects this in the fact that there are no valid signatures in the lower 6 rows corresponding to non-canonical encodings of R.

**Comment by romen** [comment]
I am not entirely sure this is a defect: we are compliant with the original `ref10` implementation, as many other libraries, that predates the changes in specs done successively to accommodate the use cases (mostly related to crypto currencies) that brought stricter requirements for the consensus problem and the changed threat model.

OpenSSL is not well suited for those applications, as it does not support batch verification, so my expectation is that it does not really affect our users or potential users.

The blog suggests "backward compatible" criteria, but it is a bit misleading as to adhere to that we would start rejecting some corner cases that we currently treat as valid, which would be a change in behavior and it's not entirely clear what are the gains for the set of use cases possible in OpenSSL.

That being said, it's not like having the same behavior everywhere would be a negative thing, but changing this behavior should be a concerted effort, across most implementations and regulators, which should start by cleaning up the mess of active specs that use different criteria, leave implementors the choice between incompatible behaviors, and in general using the same name "Ed25519" for things that behave slightly differently depending on the catered use case.
Unfortunately the blog solution of adding yet another spec, specific to the project, doesn't seem to go toward a concerted solution!
Other projects listed in the blog seems to be either forking a special-purpose subproject to cater to the consensus use cases, or in general avoiding the problem of being interoperable with other projects (or even with other versions of the same project).

**TL;DR**: fixing this without coordination with other major implementations and without gathering enough momentum among regulators and users to actually address the root of the problem (i.e., the mess of incompatible specs), there is little to gain and an uncomfortable risk level into "fixing" this in OpenSSL in isolation.

**Comment by romen** [comment]
@t-j-h the problem is that it's not just 2 options!

As the blog post highlights there is a number of items that each require a choice of being strict or "forgiving" (and again the forgiving version is not just "for compatibility" but was the one intentionally argued for in the original design of the algorithm).

If there was a unified vision on what should be the "strict" standard, then yes, we could have a solution like the one you suggest, where we have the option between a "compatibility" variant and a "strict" variant, and decide which one is default and which one is opt-in in 1.1.1 and 3.0, but we need to reach that point in coordination with the other players otherwise this just becomes a mess were we need to have "compat0", "compat1", "compat2", etc. options (and no standardized naming to let informed people decide among them!)....

**Comment by t-j-h** [comment]
This context indicates to me that we should have an option for handling this one way or the other way and what you are arguing about is what the default value of that option should be - which is separate from being able to support **both** approaches.

**Comment by t8m** [comment]
Interestingly for some reason I see @t-j-h comment after the @romen 's reaction to it. Some spacetime continuum error? :grin:

**Comment by nhorman** [comment]
It seems no such momentum was ever, or is likely to be generated to change this

Marking as inactive, to be closed at the end of the 3.4 dev cycle, barring further input

**Comment by t8m** [comment]
Closing as the issue was labeled as inactive for extended period of time and there was no additional input.
Please add a comment if you have any further information to support reopening the issue.
